pkg/config/parameters.go

```diff
@@ -328,6 +329,35 @@ func (t TimeoutParameters) Validate() error {
 	return nil
 }
 
+type HeaderFields []string
```

Comment:
Could we do something like https://github.com/projectcontour/contour/blob/0829d51420663511f0c7de3d0beacc753001605c/apis/projectcontour/v1/httpproxy.go#L615-L625

or even `map[string]string` instead so we don't have to do splitting etc.

Would also reduce duplication we have here: https://github.com/projectcontour/contour/pull/3270/files#diff-1e10012fb1102e991322c6675c2ce495873c971f5c957ea170e2cf95640f8157R274

Reply:
Looking at the existing values in the config file, I was thinking that:
```
set-request-headers:
- X-Envoy-Hostname=%HOSTNAME%
- l5d-dst-override=%CONTOUR_SERVICE_NAME%.%CONTOUR_NAMESPACE%.svc.cluster.local:%CONTOUR_SERVICE_PORT%
```
would be more natural than:
```
set-request-headers:
- name: X-Envoy-Hostname
  value: %HOSTNAME%
- name: l5d-dst-override
  value: %CONTOUR_SERVICE_NAME%.%CONTOUR_NAMESPACE%.svc.cluster.local:%CONTOUR_SERVICE_PORT%
```
Happy to change this if you think otherwise though.

Reply:
Thinking some more about your comment about reducing duplication, perhaps we should do what you suggest but take it further and use:
```
defaults:
  requestHeadersPolicy:
    set:
    - name: X-Envoy-Hostname
      value: %HOSTNAME%
    - name: l5d-dst-override
      value: %CONTOUR_SERVICE_NAME%.%CONTOUR_NAMESPACE%.svc.cluster.local:%CONTOUR_SERVICE_PORT%
    remove:
    - Some-Header
    - Some-Other-Header
```
so configuration is the same as in the object itself.

My next proposal is going to be to add a default authorisation policy which could follow a similar pattern and reuse the existing objects:
```
defaults:
  authorization:
    extensionRef:
      name: htpasswd
      namespace: projectcontour-auth
  requestHeadersPolicy:
    set:
    - name: X-Envoy-Hostname
      value: %HOSTNAME%
    - name: l5d-dst-override
      value: %CONTOUR_SERVICE_NAME%.%CONTOUR_NAMESPACE%.svc.cluster.local:%CONTOUR_SERVICE_PORT%
```
In future I could see this being extended to rateLimitPolicy, timeoutPolicy, retryPolicy etc if requested by users.

Thoughts?

Reply:
that could be interesting, though I'm not sure our philosophy on the config file having a dependency on types from the CRDs/APIs, I could see some unfortunate migration issues longer term and we've also used the config file as a less strictly versioned surface area

cc @projectcontour/maintainers for more thoughts

Reply:
I think it makes sense to keep the config file package separate from the CRD config so any tool generating configs doesn't pull in the CRD packages as well

Reply:
Yes, I agree that minimising dependencies between packages is good, though we could duplicate the types for now, which I think is what you were proposing above for HeaderValue?

Reply:
yep 👍🏽

Reply:
Ok great, I'll rework the changes to use duplicates of the types.  And add more tests.

Reply:
The contour/internal/dag package already has duplicate types so presumably those could be used since they're internal?

Reply:
unfortunately though it is a lot of duplication we would have to copy that type into the `pkg/config` package, `internal` can't be imported by code outside of `github.com/projectcontour/contour` so tools that would look to build contour config (e.g. contour-operator) wouldn't be able to use it